### PR TITLE
[misc] [doc] [frontend] LLM torch profiler support

### PR DESCRIPTION
**Pull request by SolitaryThinker**
follow up to #7451 
cc @comaniac @DamonFool 

thanks for the tips from @sfc-gh-mkeralapura

### CODE REVIEW

File: vllm/executor/gpu_executor.py [hunk 1]
```diff
@@ -168,6 +168,12 @@ def check_health(self) -> None:
         # it's running.
         return
 
+    def start_profile(self) -> None:
```
Comment (DamonFool) [inline vllm/executor/gpu_executor.py:171]: Maybe, we should also add start_profile/stop_profile for CPU-only targets?

Comment (ywang96) [reply vllm/executor/gpu_executor.py:171]: +1, either that or make it clear in the documentation & example that this is currently only supported on GPUs

Comment (SolitaryThinker) [reply vllm/executor/gpu_executor.py:172]: Thanks for the suggestion, missed it. @DamonFool @ywang96 added to `cpu_executor.py` in [2b23e8f](https://github.com/vllm-project/vllm/pull/7943/commits/2b23e8f3dfc4a359ab62e47c009b393dcc7a8f56) PTAL.

**Review by comaniac — APPROVED** [review]
Somehow missed this PR before. LGTM

**Review by DamonFool — APPROVED** [review]
Thanks for the update.
LGTM

**Comment by sylviayangyy** [comment]
not sure if it's appropriate to ask this here, just in case if there is any help: when I tried to run examples/offline_inference_with_profiler.py, occuring the following error `RuntimeError: !stack.empty() INTERNAL ASSERT FAILED at "../torch/csrc/autograd/profiler_python.cpp":969, please report a bug to PyTorch. Python replay stack is empty`, the only change I do to the code is changing the facebook/opt-125m model to my local model Llama-2-7b-hf

<details>
<summary>The output of `python collect_env.py`</summary>

```text
Collecting environment information...
PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.12.4 | packaged by Anaconda, Inc. | (main, Jun 18 2024, 15:12:24) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.56.bsk.9-amd64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A10
GPU 1: NVIDIA A10
GPU 2: NVIDIA A10
GPU 3: NVIDIA A10

Nvidia driver version: 470.129.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 57 bits virtual
Byte Order:                      Little Endian
CPU(s):                          128
On-line CPU(s) list:             0-127
Vendor ID:                       GenuineIntel
BIOS Vendor ID:                  Intel(R) Corporation
Model name:                      Intel(R) Xeon(R) Platinum 8336C CPU @ 2.30GHz
BIOS Model name:                 Intel(R) Xeon(R) Platinum 8336C CPU @ 2.30GHz
CPU family:                      6
Model:                           106
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
Stepping:                        6
CPU max MHz:                     3500.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4600.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       3 MiB (64 instances)
L1i cache:                       2 MiB (64 instances)
L2 cache:                        80 MiB (64 instances)
L3 cache:                        108 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-31,64-95
NUMA node1 CPU(s):               32-63,96-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-ml-py==12.560.30
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] nvidia-nvjitlink-cu12==12.6.20
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] pyzmq==26.1.1
[pip3] torch==2.4.0
[pip3] torch-tb-profiler==0.4.3
[pip3] torchvision==0.19.0
[pip3] transformers==4.44.1
[pip3] triton==3.0.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
[conda] nvidia-ml-py              12.560.30                pypi_0    pypi
[conda] nvidia-nccl-cu12          2.20.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.6.20                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
[conda] pyzmq                     26.1.1                   pypi_0    pypi
[conda] torch                     2.4.0                    pypi_0    pypi
[conda] torch-tb-profiler         0.4.3                    pypi_0    pypi
[conda] torchvision               0.19.0                   pypi_0    pypi
[conda] transformers              4.44.1                   pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: N/A
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0	GPU1	GPU2	GPU3	mlx5_0	mlx5_1	CPU Affinity	NUMA Affinity
GPU0	 X 	NODE	SYS	SYS	NODE	SYS	0-31,64-95	0
GPU1	NODE	 X 	SYS	SYS	NODE	SYS	0-31,64-95	0
GPU2	SYS	SYS	 X 	NODE	SYS	NODE	32-63,96-127	1
GPU3	SYS	SYS	NODE	 X 	SYS	NODE	32-63,96-127	1
mlx5_0	NODE	NODE	SYS	SYS	 X 	SYS
mlx5_1	SYS	SYS	NODE	NODE	SYS	 X

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks
```

</details>

<details>
<summary>The error stack</summary>

```text
INFO 09-17 11:46:30 model_runner.py:1428] Graph capturing finished in 9 secs.
Processed prompts: 100%|██████████████████████████████████████████████| 4/4 [00:00<00:00,  6.80it/s, est. speed input: 45.93 toks/s, output: 108.88 toks/s]
[rank0]: Traceback (most recent call last):
[rank0]:   File "/root/vllm/examples/offline_inference_with_profiler_local_model.py", line 27, in <module>
[rank0]:     llm.stop_profile()
[rank0]:   File "/root/vllm/vllm/entrypoints/llm.py", line 577, in stop_profile
[rank0]:     self.llm_engine.stop_profile()
[rank0]:   File "/root/vllm/vllm/engine/llm_engine.py", line 1612, in stop_profile
[rank0]:     self.model_executor.stop_profile()
[rank0]:   File "/root/vllm/vllm/executor/gpu_executor.py", line 176, in stop_profile
[rank0]:     self.driver_worker.stop_profile()
[rank0]:   File "/root/vllm/vllm/worker/worker.py", line 145, in stop_profile
[rank0]:     self.profiler.stop()
[rank0]:   File "/root/miniconda3/lib/python3.12/site-packages/torch/profiler/profiler.py", line 722, in stop
[rank0]:     self._transit_action(self.current_action, None)
[rank0]:   File "/root/miniconda3/lib/python3.12/site-packages/torch/profiler/profiler.py", line 751, in _transit_action
[rank0]:     action()
[rank0]:   File "/root/miniconda3/lib/python3.12/site-packages/torch/profiler/profiler.py", line 206, in stop_trace
[rank0]:     self.profiler.__exit__(None, None, None)
[rank0]:   File "/root/miniconda3/lib/python3.12/site-packages/torch/autograd/profiler.py", line 352, in __exit__
[rank0]:     self.kineto_results = _disable_profiler()
[rank0]:                           ^^^^^^^^^^^^^^^^^^^
[rank0]: RuntimeError: !stack.empty() INTERNAL ASSERT FAILED at "../torch/csrc/autograd/profiler_python.cpp":969, please report a bug to PyTorch. Python replay stack is empty.
```

</details>

<details>
<summary>my code</summary>

```code
import os

from vllm import LLM, SamplingParams

# enable torch profiler, can also be set on cmd line
os.environ["VLLM_TORCH_PROFILER_DIR"] = "./vllm_profile"

# Sample prompts.
prompts = [
    "Hello, my name is",
    "The president of the United States is",
    "The capital of France is",
    "The future of AI is",
]
# Create a sampling params object.
sampling_params = SamplingParams(temperature=0.8, top_p=0.95)

# Create an LLM.
llm = LLM(model="/root/Llama-2-7b-hf", tensor_parallel_size=1)

llm.start_profile()

# Generate texts from the prompts. The output is a list of RequestOutput objects
# that contain the prompt, generated text, and other information.
outputs = llm.generate(prompts, sampling_params)

llm.stop_profile()

# Print the outputs.
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
```

</details>

**Comment by sylviayangyy** [comment]
Besides, do you think if it's a good idea to create a specific configuration for profiling? In my use case, I care about memory usage too, so I add `profile_memory=True` option when initializing the profiler like this:
image
However, since PyTorch Profiler offers many more options, I’m considering whether it would be beneficial to abstract this into a `ProfileConfig`. This way, vllm users could easily and flexibly customize their profiling according to different needs. cc @SolitaryThinker @comaniac @DamonFool

**Comment by comaniac** [comment]
Feel free to open an issue for bug report and discussions like this.

**Comment by sylviayangyy** [comment]
> Feel free to open an issue for bug report and discussions like this.

got it, will do, thank you!

**Pull request merged**